zipkin-collector/scribe/src/main/java/zipkin2/collector/scribe/ScribeInboundHandler.java

```diff
@@ -0,0 +1,184 @@
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufHolder;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ScribeInboundHandler extends ChannelInboundHandlerAdapter {
+
+  static final Logger logger = LoggerFactory.getLogger(ScribeInboundHandler.class);
+
+  // Headers mostly copied from https://github.com/apache/thrift/blob/master/lib/javame/src/org/apache/thrift/transport/THttpClient.java#L130
+  static final HttpHeaders THRIFT_HEADERS = HttpHeaders.of(
+    HttpMethod.POST, "/internal/zipkin-thriftrpc")
+    .set(HttpHeaderNames.CONTENT_TYPE, "application/x-thrift")
+    .set(HttpHeaderNames.ACCEPT, "application/x-thrift")
+    .set(HttpHeaderNames.USER_AGENT, "Zipkin/ScribeInboundHandler")
+    .asImmutable();
+
+  final THttpService scribeService;
+
+  ScribeInboundHandler(ScribeSpanConsumer scribe) {
+    scribeService = THttpService.of(scribe);
+  }
+
+  enum ReadState {
+    HEADER,
+    PAYLOAD
+  }
+
+  CompositeByteBuf pending;
+  ReadState state;
+  int nextFrameSize;
+
+  @Override
+  public void channelActive(ChannelHandlerContext ctx) throws Exception {
+    pending = ctx.alloc().compositeBuffer();
+    state = ReadState.HEADER;
+  }
+
+  @Override
+  public void channelRead(final ChannelHandlerContext ctx, Object msg) {
+    if (pending == null) {
+      // Already closed (probably due to an exception).
+      return;
+    }
+
+    assert msg instanceof ByteBuf;
+    ByteBuf buf = (ByteBuf) msg;
+    pending.addComponent(true, buf);
+
+    switch (state) {
+      case HEADER:
+        maybeReadHeader(ctx);
+        break;
+      case PAYLOAD:
+        maybeReadPayload(ctx);
+        break;
+    }
+  }
+
+  @Override
+  public void channelInactive(ChannelHandlerContext ctx) {
+    release();
+  }
+
+  @Override
+  public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+    logger.warn("Unexpected error handling connection.", cause);
+
+    release();
+    closeOnFlush(ctx.channel());
+  }
+
+  void maybeReadHeader(ChannelHandlerContext ctx) {
+    if (pending.readableBytes() < 4) {
+      return;
+    }
+    nextFrameSize = pending.readInt();
+    state = ReadState.PAYLOAD;
+    maybeReadPayload(ctx);
+  }
+
+  void maybeReadPayload(ChannelHandlerContext ctx) {
+    if (pending.readableBytes() < nextFrameSize) {
+      return;
+    }
+
+    ByteBuf payload = ctx.alloc().buffer(nextFrameSize);
+    pending.readBytes(payload, nextFrameSize);
+    pending.discardSomeReadBytes();
+
+    state = ReadState.HEADER;
+
+    HttpRequest request = HttpRequest.of(
+      THRIFT_HEADERS.toMutable(),
+      new ByteBufHttpData(payload, true));
+    ServiceRequestContext requestContext = ServiceRequestContextBuilder.of(request)
```

Comment:
@trustin Can you confirm this code looks like what you'd expect for invoking a service manually?